File: python/cadreen/resources/learning.py
from __future__ import annotations

from ..client import HttpClient
from ..types import (
    ListLearningPatternsResponse,
    ListLearningEpisodesResponse,
    ListLearningSuggestionsResponse,
    LearningPattern,
    LearningEpisode,
    LearningSuggestion,
    Pagination,
)


class LearningResource:
    def __init__(self, client: HttpClient) -> None:
        self._client = client

    async def patterns(self) -> ListLearningPatternsResponse:
        raw = await self._client.get("/api/v1/cadreen/learning/patterns")
        patterns = [
            LearningPattern(
                id=p["id"],
                pattern=p["pattern"],
                confidence=p.get("confidence", 0.0),
                occurrences=p.get("occurrences"),
                domain=p.get("domain"),
                tags=p.get("tags"),
                created_at=p.get("created_at"),
            )
            for p in raw.get("patterns", [])
        ]
        pagination = None
        if raw.get("pagination"):
            p = raw["pagination"]
            pagination = Pagination(limit=p["limit"], offset=p["offset"], has_more=p["has_more"])
        return ListLearningPatternsResponse(
            patterns=patterns,
            count=raw.get("count", 0),
            pagination=pagination,
        )

    async def episodes(self) -> ListLearningEpisodesResponse:
        raw = await self._client.get("/api/v1/cadreen/learning/episodes")
        episodes = [
            LearningEpisode(
                id=e["id"],
                description=e["description"],
                outcome=e.get("outcome"),
                trace_id=e.get("trace_id"),
                domain=e.get("domain"),
                created_at=e.get("created_at"),
            )
            for e in raw.get("episodes", [])
        ]
        pagination = None
        if raw.get("pagination"):
            p = raw["pagination"]
            pagination = Pagination(limit=p["limit"], offset=p["offset"], has_more=p["has_more"])
        return ListLearningEpisodesResponse(
            episodes=episodes,
            count=raw.get("count", 0),
            pagination=pagination,
        )

    async def suggestions(self) -> ListLearningSuggestionsResponse:
        raw = await self._client.get("/api/v1/cadreen/learning/suggestions")
        suggestions = [
            LearningSuggestion(
                id=s["id"],
                type=s["type"],
                description=s["description"],
                impact=s.get("impact"),
                domain=s.get("domain"),
            )
            for s in raw.get("suggestions", [])
        ]
        pagination = None
        if raw.get("pagination"):
            p = raw["pagination"]
            pagination = Pagination(limit=p["limit"], offset=p["offset"], has_more=p["has_more"])
        return ListLearningSuggestionsResponse(
            suggestions=suggestions,
            count=raw.get("count", 0),
            pagination=pagination,
        )

File: go/cmd/cadreen/commands/learning.go
package commands

import (
	"encoding/json"
	"fmt"

	"github.com/timothy-billingrails/cadreen-sdks/go/cmd/cadreen/output"

	"github.com/spf13/cobra"
)

var learningCmd = &cobra.Command{
	Use:   "learning",
	Short: "Learning insights (patterns, episodes, suggestions)",
	Long: `View detected patterns, learning episodes, and improvement suggestions.

Examples:
  cadreen learning patterns
  cadreen learning episodes
  cadreen learning suggestions`,
}

var learningPatternsCmd = &cobra.Command{
	Use:   "patterns",
	Short: "Detected patterns",
	RunE: func(cmd *cobra.Command, args []string) error {
		requireAuth()

		resp, err := rawDo("GET", "/api/v1/cadreen/learning/patterns", nil)
		if err != nil {
			return handleAPIError(err)
		}
		defer resp.Body.Close()

		if resp.StatusCode != 200 {
			return handleHTTPError(resp)
		}

		var result struct {
			Patterns []struct {
				ID          string   `json:"id"`
				Pattern     string   `json:"pattern"`
				Confidence  float64  `json:"confidence"`
				Occurrences int      `json:"occurrences"`
				Domain      string   `json:"domain"`
			} `json:"patterns"`
			Count int `json:"count"`
		}
		if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
			return err
		}

		format := outputFormat()
		switch format {
		case output.FormatJSON:
			output.Print(result, format)
		default:
			if len(result.Patterns) == 0 {
				fmt.Println("No patterns detected yet.")
				return nil
			}
			fmt.Printf("Detected patterns (%d):\n\n", result.Count)
			for _, p := range result.Patterns {
				fmt.Printf("  %s\n", p.Pattern)
				fmt.Printf("    Confidence: %.0f%%", p.Confidence*100)
				if p.Domain != "" {
					fmt.Printf("  Domain: %s", p.Domain)
				}
				fmt.Println()
			}
		}

		return nil
	},
}

var learningEpisodesCmd = &cobra.Command{
	Use:   "episodes",
	Short: "Learning episodes",
	RunE: func(cmd *cobra.Command, args []string) error {
		requireAuth()

		resp, err := rawDo("GET", "/api/v1/cadreen/learning/episodes", nil)
		if err != nil {
			return handleAPIError(err)
		}
		defer resp.Body.Close()

		if resp.StatusCode != 200 {
			return handleHTTPError(resp)
		}

		var result struct {
			Episodes []struct {
				ID          string `json:"id"`
				Description string `json:"description"`
				Outcome     string `json:"outcome"`
				Domain      string `json:"domain"`
			} `json:"episodes"`
			Count int `json:"count"`
		}
		if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
			return err
		}

		format := outputFormat()
		switch format {
		case output.FormatJSON:
			output.Print(result, format)
		default:
			if len(result.Episodes) == 0 {
				fmt.Println("No learning episodes yet.")
				return nil
			}
			fmt.Printf("Learning episodes (%d):\n\n", result.Count)
			for _, e := range result.Episodes {
				fmt.Printf("  %s\n", e.Description)
				if e.Outcome != "" {
					fmt.Printf("    Outcome: %s\n", e.Outcome)
				}
				fmt.Println()
			}
		}

		return nil
	},
}

var learningSuggestionsCmd = &cobra.Command{
	Use:   "suggestions",
	Short: "Improvement suggestions",
	RunE: func(cmd *cobra.Command, args []string) error {
		requireAuth()

		resp, err := rawDo("GET", "/api/v1/cadreen/learning/suggestions", nil)
		if err != nil {
			return handleAPIError(err)
		}
		defer resp.Body.Close()

		if resp.StatusCode != 200 {
			return handleHTTPError(resp)
		}

		var result struct {
			Suggestions []struct {
				ID          string `json:"id"`
				Type        string `json:"type"`
				Description string `json:"description"`
				Impact      string `json:"impact"`
			} `json:"suggestions"`
			Count int `json:"count"`
		}
		if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
			return err
		}

		format := outputFormat()
		switch format {
		case output.FormatJSON:
			output.Print(result, format)
		default:
			if len(result.Suggestions) == 0 {
				fmt.Println("No suggestions at this time.")
				return nil
			}
			fmt.Printf("Suggestions (%d):\n\n", result.Count)
			for _, s := range result.Suggestions {
				fmt.Printf("  [%s] %s\n", s.Type, s.Description)
				if s.Impact != "" {
					fmt.Printf("    Impact: %s\n", s.Impact)
				}
				fmt.Println()
			}
		}

		return nil
	},
}

func init() {
	learningCmd.AddCommand(learningPatternsCmd)
	learningCmd.AddCommand(learningEpisodesCmd)
	learningCmd.AddCommand(learningSuggestionsCmd)
	rootCmd.AddCommand(learningCmd)
}
